File: src/engine/ContextEngineImpl.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  AggregatedAIContext,
  ContextDiff,
  ContextScope,
  ContextWatcher,
  IContextEngine,
  MemorySlot,
  TokenBudgetConfig,
} from './types';

export class ContextEngineImpl implements IContextEngine {
  private scopes: Map<string, ContextScope> = new Map();
  private rootScopeId: string = 'global';
  private watchers: Map<string, { pattern: string; callback: ContextWatcher; scopeId?: string }> = new Map();
  private diffHistory: ContextDiff[] = [];
  private maxDiffHistorySize: number;
  private isTransactionActive: boolean = false;
  private transactionBackup?: Map<string, ContextScope>;

  constructor(options?: { maxDiffHistorySize?: number }) {
    this.maxDiffHistorySize = options?.maxDiffHistorySize ?? 500;
    this.initRootScope();
  }

  private initRootScope(): void {
    const root: ContextScope = {
      id: this.rootScopeId,
      name: 'Global Root Scope',
      variables: {
        system: {
          version: '1.0.0',
          environment: 'production',
          status: 'HEALTHY',
          uptime: Date.now(),
        },
        session: {
          id: `ses_${Date.now()}`,
          activeUser: 'sauljtm25@gmail.com',
          role: 'ADMIN',
        },
        appState: {
          mode: 'REACTIVE_ORCHESTRATION',
          activeWorkflows: 0,
        },
      },
      memorySlots: {},
      metadata: {
        createdAt: Date.now(),
        updatedAt: Date.now(),
        version: 1,
        description: 'Default top-level system context',
      },
    };
    this.scopes.set(this.rootScopeId, root);
  }

  public createScope(name: string, parentScopeId?: string, initialData?: Record<string, any>): ContextScope {
    const parentId = parentScopeId || this.rootScopeId;
    const scopeId = `scope_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;

    const scope: ContextScope = {
      id: scopeId,
      name,
      parentScopeId: parentId,
      variables: initialData ? JSON.parse(JSON.stringify(initialData)) : {},
      memorySlots: {},
      metadata: {
        createdAt: Date.now(),
        updatedAt: Date.now(),
        version: 1,
      },
    };

    this.scopes.set(scopeId, scope);

    this.recordDiff({
      timestamp: Date.now(),
      scopeId,
      path: '$scope',
      oldValue: null,
      newValue: { id: scopeId, name, parentScopeId: parentId },
      operation: 'CREATE_SCOPE',
    });

    return scope;
  }

  public getScope(scopeId: string = this.rootScopeId): ContextScope | undefined {
    return this.scopes.get(scopeId);
  }

  public getAllScopes(): ContextScope[] {
    return Array.from(this.scopes.values());
  }

  public deleteScope(scopeId: string): boolean {
    if (scopeId === this.rootScopeId) {
      console.warn('Cannot delete root global context scope');
      return false;
    }
    const existed = this.scopes.delete(scopeId);
    if (existed) {
      this.recordDiff({
        timestamp: Date.now(),
        scopeId,
        path: '$scope',
        oldValue: scopeId,
        newValue: null,
        operation: 'DELETE_SCOPE',
      });
    }
    return existed;
  }

  public get<T = any>(path: string, scopeId: string = this.rootScopeId): T | undefined {
    let currentScope = this.scopes.get(scopeId);

    // Look up hierarchically
    while (currentScope) {
      const val = this.getByPath(currentScope.variables, path);
      if (val !== undefined) {
        return val as T;
      }
      if (currentScope.parentScopeId) {
        currentScope = this.scopes.get(currentScope.parentScopeId);
      } else {
        break;
      }
    }

    return undefined;
  }

  public set<T = any>(path: string, value: T, scopeId: string = this.rootScopeId, sourceEventId?: string): void {
    const scope = this.scopes.get(scopeId);
    if (!scope) {
      throw new Error(`Context scope '${scopeId}' not found`);
    }

    const oldValue = this.getByPath(scope.variables, path);
    this.setByPath(scope.variables, path, value);
    scope.metadata.updatedAt = Date.now();
    scope.metadata.version++;

    this.recordDiff({
      timestamp: Date.now(),
      scopeId,
      path,
      oldValue,
      newValue: value,
      operation: 'SET',
      sourceEventId,
    });
  }

  public patch(
    path: string,
    partialValue: Record<string, any>,
    scopeId: string = this.rootScopeId,
    sourceEventId?: string
  ): void {
    const scope = this.scopes.get(scopeId);
    if (!scope) {
      throw new Error(`Context scope '${scopeId}' not found`);
    }

    const current = this.getByPath(scope.variables, path) || {};
    const updated = typeof current === 'object' && !Array.isArray(current)
      ? { ...current, ...partialValue }
      : partialValue;

    const oldValue = current;
    this.setByPath(scope.variables, path, updated);
    scope.metadata.updatedAt = Date.now();
    scope.metadata.version++;

    this.recordDiff({
      timestamp: Date.now(),
      scopeId,
      path,
      oldValue,
      newValue: updated,
      operation: 'PATCH',
      sourceEventId,
    });
  }

  public delete(path: string, scopeId: string = this.rootScopeId): boolean {
    const scope = this.scopes.get(scopeId);
    if (!scope) return false;

    const oldValue = this.getByPath(scope.variables, path);
    if (oldValue === undefined) return false;

    this.deleteByPath(scope.variables, path);
    scope.metadata.updatedAt = Date.now();
    scope.metadata.version++;

    this.recordDiff({
      timestamp: Date.now(),
      scopeId,
      path,
      oldValue,
      newValue: undefined,
      operation: 'DELETE',
    });

    return true;
  }

  public watch(pathPattern: string, callback: ContextWatcher, scopeId?: string): () => void {
    const watcherId = `watcher_${Date.now()}_${Math.random().toString(36).substring(2, 7)}`;
    this.watchers.set(watcherId, { pattern: pathPattern, callback, scopeId });

    return () => {
      this.watchers.delete(watcherId);
    };
  }

  public setMemorySlot(
    key: string,
    value: any,
    options?: {
      scopeId?: string;
      importance?: number;
      tags?: string[];
      ttlMs?: number;
      summary?: string;
      sourceEventId?: string;
    }
  ): MemorySlot {
    const scopeId = options?.scopeId || this.rootScopeId;
    const scope = this.scopes.get(scopeId);
    if (!scope) {
      throw new Error(`Scope '${scopeId}' not found`);
    }

    const now = Date.now();
    const slot: MemorySlot = {
      id: `mem_${now}_${Math.random().toString(36).substring(2, 7)}`,
      key,
      value,
      importance: options?.importance ?? 5,
      tags: options?.tags ?? [],
      createdAt: now,
      updatedAt: now,
      expiresAt: options?.ttlMs ? now + options.ttlMs : undefined,
      summary: options?.summary,
      sourceEventId: options?.sourceEventId,
    };

    scope.memorySlots[key] = slot;
    scope.metadata.updatedAt = now;
    scope.metadata.version++;

    this.recordDiff({
      timestamp: now,
      scopeId,
      path: `memorySlots.${key}`,
      oldValue: null,
      newValue: slot,
      operation: 'SET',
      sourceEventId: options?.sourceEventId,
    });

    return slot;
  }

  public getMemorySlots(scopeId: string = this.rootScopeId, minImportance: number = 1): MemorySlot[] {
    const scope = this.scopes.get(scopeId);
    if (!scope) return [];

    const now = Date.now();
    const result: MemorySlot[] = [];

    for (const [key, slot] of Object.entries(scope.memorySlots)) {
      if (slot.expiresAt && slot.expiresAt <= now) {
        delete scope.memorySlots[key];
        continue;
      }
      if (slot.importance >= minImportance) {
        result.push(slot);
      }
    }

    return result.sort((a, b) => b.importance - a.importance || b.updatedAt - a.updatedAt);
  }

  public deleteMemorySlot(key: string, scopeId: string = this.rootScopeId): boolean {
    const scope = this.scopes.get(scopeId);
    if (!scope || !scope.memorySlots[key]) return false;

    delete scope.memorySlots[key];
    scope.metadata.updatedAt = Date.now();

    this.recordDiff({
      timestamp: Date.now(),
      scopeId,
      path: `memorySlots.${key}`,
      oldValue: key,
      newValue: undefined,
      operation: 'DELETE',
    });

    return true;
  }

  public getSnapshot(scopeId?: string): Record<string, any> {
    if (scopeId) {
      const s = this.scopes.get(scopeId);
      return s ? JSON.parse(JSON.stringify(s)) : {};
    }

    const allSnapshots: Record<string, any> = {};
    for (const [id, s] of this.scopes.entries()) {
      allSnapshots[id] = JSON.parse(JSON.stringify(s));
    }
    return allSnapshots;
  }

  public restoreSnapshot(snapshot: Record<string, any>, scopeId?: string): void {
    if (scopeId && snapshot.id) {
      this.scopes.set(scopeId, JSON.parse(JSON.stringify(snapshot)));
      return;
    }

    this.scopes.clear();
    for (const [id, s] of Object.entries(snapshot)) {
      this.scopes.set(id, JSON.parse(JSON.stringify(s)));
    }

    if (!this.scopes.has(this.rootScopeId)) {
      this.initRootScope();
    }
  }

  public getDiffHistory(limit: number = 50): ContextDiff[] {
    return this.diffHistory.slice(-limit);
  }

  /**
   * Generates token-budgeted, structured context ready for LLM Prompt injection
   */
  public aggregateContextForAI(config?: TokenBudgetConfig, targetScopeId?: string): AggregatedAIContext {
    const scopeId = targetScopeId || this.rootScopeId;
    const hierarchy: string[] = [];
    let current = this.scopes.get(scopeId);

    const mergedVariables: Record<string, any> = {};
    const memoriesMap: Map<string, MemorySlot> = new Map();

    // Traverse upwards to inherit global parent data
    const pathScopes: ContextScope[] = [];
    while (current) {
      pathScopes.unshift(current);
      if (current.parentScopeId) {
        current = this.scopes.get(current.parentScopeId);
      } else {
        break;
      }
    }

    for (const s of pathScopes) {
      hierarchy.push(`${s.name} [${s.id}]`);
      Object.assign(mergedVariables, JSON.parse(JSON.stringify(s.variables)));
      for (const slot of this.getMemorySlots(s.id, config?.minImportanceThreshold ?? 3)) {
        memoriesMap.set(slot.key, slot);
      }
    }

    const relevantMemories = Array.from(memoriesMap.values()).sort(
      (a, b) => b.importance - a.importance || b.updatedAt - a.updatedAt
    );

    // Create system prompt block
    const sysPrompt = [
      '### SYSTEM CONTEXT & ACTIVE STATE ###',
      `Hierarchy: ${hierarchy.join(' -> ')}`,
      `Active Context Variables: ${JSON.stringify(mergedVariables, null, 2)}`,
      relevantMemories.length > 0
        ? `Long-Term/Working Memory Slots:\n${relevantMemories
            .map((m) => `- [Importance ${m.importance}/10] ${m.key}: ${JSON.stringify(m.value)} ${m.summary ? `(${m.summary})` : ''}`)
            .join('\n')}`
        : 'Memory: None',
      '#####################################',
    ].join('\n\n');

    // Rough token estimator (~4 chars per token)
    const estimatedTokens = Math.ceil(sysPrompt.length / 4);

    return {
      systemPromptAddition: sysPrompt,
      structuredContext: mergedVariables,
      relevantMemories,
      estimatedTokens,
      scopeHierarchy: hierarchy,
      generatedAt: Date.now(),
    };
  }

  public transaction<R>(fn: (ctx: IContextEngine) => R, _scopeId?: string): R {
    if (this.isTransactionActive) {
      return fn(this);
    }

    this.isTransactionActive = true;
    this.transactionBackup = new Map();
    for (const [k, v] of this.scopes.entries()) {
      this.transactionBackup.set(k, JSON.parse(JSON.stringify(v)));
    }

    try {
      const result = fn(this);
      this.isTransactionActive = false;
      this.transactionBackup = undefined;
      return result;
    } catch (error) {
      // Rollback
      this.scopes = this.transactionBackup!;
      this.isTransactionActive = false;
      this.transactionBackup = undefined;
      throw error;
    }
  }

  public reset(): void {
    this.scopes.clear();
    this.diffHistory = [];
    this.initRootScope();
  }

  // --------------------------------------------------------------------------
  // Internal Helpers
  // --------------------------------------------------------------------------

  private recordDiff(diff: ContextDiff): void {
    this.diffHistory.push(diff);
    if (this.diffHistory.length > this.maxDiffHistorySize) {
      this.diffHistory.shift();
    }

    // Trigger matching watchers
    for (const watcher of this.watchers.values()) {
      if (watcher.scopeId && watcher.scopeId !== diff.scopeId) {
        continue;
      }

      if (this.matchPath(watcher.pattern, diff.path)) {
        const scope = this.scopes.get(diff.scopeId);
        if (scope) {
          try {
            watcher.callback(diff, scope);
          } catch (e) {
            console.error('Error in context watcher callback:', e);
          }
        }
      }
    }
  }

  private matchPath(pattern: string, path: string): boolean {
    if (pattern === '*' || pattern === '**' || pattern === path) return true;
    if (pattern.endsWith('.*')) {
      const prefix = pattern.slice(0, -2);
      return path.startsWith(prefix);
    }
    return false;
  }

  private getByPath(obj: any, path: string): any {
    if (!obj || typeof obj !== 'object') return undefined;
    const parts = path.split('.');
    let cur = obj;
    for (const part of parts) {
      if (cur === null || cur === undefined) return undefined;
      cur = cur[part];
    }
    return cur;
  }

  private setByPath(obj: any, path: string, value: any): void {
    const parts = path.split('.');
    let cur = obj;
    for (let i = 0; i < parts.length - 1; i++) {
      const part = parts[i];
      if (!(part in cur) || typeof cur[part] !== 'object' || cur[part] === null) {
        cur[part] = {};
      }
      cur = cur[part];
    }
    cur[parts[parts.length - 1]] = value;
  }

  private deleteByPath(obj: any, path: string): void {
    const parts = path.split('.');
    let cur = obj;
    for (let i = 0; i < parts.length - 1; i++) {
      const part = parts[i];
      if (!cur[part] || typeof cur[part] !== 'object') return;
      cur = cur[part];
    }
    delete cur[parts[parts.length - 1]];
  }
}

File: android-clean-architecture/src/main/kotlin/com/uru/data/autonomy/ContextEngineImpl.kt
package com.autonomy.engine.data

import com.autonomy.engine.domain.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.regex.Pattern

/**
 * Thread-safe hierarchical ContextEngine with memory TTL, deep paths,
 * rollback transactions, and AI context aggregation.
 */
class ContextEngineImpl : ContextEngine {

    private val scopes = ConcurrentHashMap<String, ContextScope>()
    private val watchers = ConcurrentHashMap<String, WatcherEntry>()
    private val diffHistory = CopyOnWriteArrayList<ContextDiff>()
    private val rootScopeId = "global"

    init {
        // Initialize Root Global Scope
        val globalScope = ContextScope(
            id = rootScopeId,
            name = "Global System Scope",
            metadata = ContextScopeMetadata(owner = "system", description = "Root system scope")
        )
        scopes[rootScopeId] = globalScope
    }

    override fun createScope(name: String, parentScopeId: String?, initialData: Map<String, Any?>): ContextScope {
        val parent = parentScopeId ?: rootScopeId
        val newScope = ContextScope(
            name = name,
            parentScopeId = parent,
            metadata = ContextScopeMetadata(description = "Scope for $name")
        )
        initialData.forEach { (k, v) ->
            newScope.variables[k] = v
        }
        scopes[newScope.id] = newScope

        recordDiff(
            ContextDiff(
                scopeId = newScope.id,
                path = "$",
                oldValue = null,
                newValue = initialData,
                operation = ContextOperation.CREATE_SCOPE
            )
        )
        return newScope
    }

    override fun getScope(scopeId: String): ContextScope? = scopes[scopeId]

    override fun getAllScopes(): List<ContextScope> = scopes.values.toList()

    override fun deleteScope(scopeId: String): Boolean {
        if (scopeId == rootScopeId) return false // Cannot delete root
        val removed = scopes.remove(scopeId) != null
        if (removed) {
            recordDiff(
                ContextDiff(
                    scopeId = scopeId,
                    path = "$",
                    oldValue = null,
                    newValue = null,
                    operation = ContextOperation.DELETE_SCOPE
                )
            )
        }
        return removed
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T> get(path: String, scopeId: String?): T? {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return null

        val localValue = getNestedValue(scope.variables, path)
        if (localValue != null) {
            return localValue as? T
        }

        // Inherit recursively from parent scope
        scope.parentScopeId?.let { parentId ->
            return get(path, parentId)
        }

        return null
    }

    override fun <T> set(path: String, value: T, scopeId: String?, sourceEventId: String?) {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return

        val oldValue = getNestedValue(scope.variables, path)
        setNestedValue(scope.variables, path, value)
        scope.metadata = scope.metadata.copy(
            updatedAt = System.currentTimeMillis(),
            version = scope.metadata.version + 1
        )

        val diff = ContextDiff(
            scopeId = targetScopeId,
            path = path,
            oldValue = oldValue,
            newValue = value,
            operation = ContextOperation.SET,
            sourceEventId = sourceEventId
        )
        recordDiff(diff)
        notifyWatchers(diff, scope)
    }

    @Suppress("UNCHECKED_CAST")
    override fun patch(path: String, partialValue: Map<String, Any?>, scopeId: String?, sourceEventId: String?) {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return

        val current = getNestedValue(scope.variables, path) as? Map<String, Any?> ?: emptyMap()
        val merged = current.toMutableMap().apply { putAll(partialValue) }
        setNestedValue(scope.variables, path, merged)

        val diff = ContextDiff(
            scopeId = targetScopeId,
            path = path,
            oldValue = current,
            newValue = merged,
            operation = ContextOperation.PATCH,
            sourceEventId = sourceEventId
        )
        recordDiff(diff)
        notifyWatchers(diff, scope)
    }

    override fun delete(path: String, scopeId: String?): Boolean {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return false

        val oldValue = getNestedValue(scope.variables, path) ?: return false
        val removed = deleteNestedValue(scope.variables, path)
        if (removed) {
            val diff = ContextDiff(
                scopeId = targetScopeId,
                path = path,
                oldValue = oldValue,
                newValue = null,
                operation = ContextOperation.DELETE
            )
            recordDiff(diff)
            notifyWatchers(diff, scope)
        }
        return removed
    }

    override fun watch(pathPattern: String, scopeId: String?, callback: ContextWatcher): () -> Unit {
        val entry = WatcherEntry(
            id = java.util.UUID.randomUUID().toString(),
            pathPattern = pathPattern,
            scopeId = scopeId,
            callback = callback
        )
        watchers[entry.id] = entry
        return { watchers.remove(entry.id) }
    }

    override fun setMemorySlot(
        key: String,
        value: Any?,
        scopeId: String?,
        importance: Int,
        tags: List<String>,
        ttlMs: Long?,
        summary: String?,
        sourceEventId: String?
    ): MemorySlot {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: scopes[rootScopeId]!!

        val expiresAt = ttlMs?.let { System.currentTimeMillis() + it }
        val slot = MemorySlot(
            key = key,
            value = value,
            importance = importance.coerceIn(1, 10),
            tags = tags,
            expiresAt = expiresAt,
            summary = summary,
            sourceEventId = sourceEventId
        )
        scope.memorySlots[key] = slot
        return slot
    }

    override fun getMemorySlots(scopeId: String?, minImportance: Int): List<MemorySlot> {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return emptyList()

        // Clean expired slots
        val now = System.currentTimeMillis()
        val validSlots = mutableListOf<MemorySlot>()
        scope.memorySlots.entries.removeIf { (_, slot) ->
            if (slot.expiresAt != null && slot.expiresAt < now) {
                true
            } else {
                if (slot.importance >= minImportance) {
                    validSlots.add(slot)
                }
                false
            }
        }
        return validSlots.sortedByDescending { it.importance }
    }

    override fun deleteMemorySlot(key: String, scopeId: String?): Boolean {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return false
        return scope.memorySlots.remove(key) != null
    }

    override fun getSnapshot(scopeId: String?): Map<String, Any?> {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return emptyMap()
        return deepCopyMap(scope.variables)
    }

    override fun restoreSnapshot(snapshot: Map<String, Any?>, scopeId: String?) {
        val targetScopeId = scopeId ?: rootScopeId
        val scope = scopes[targetScopeId] ?: return
        scope.variables.clear()
        snapshot.forEach { (k, v) -> scope.variables[k] = v }
    }

    override fun getDiffHistory(limit: Int): List<ContextDiff> {
        return diffHistory.takeLast(limit)
    }

    override fun aggregateContextForAI(config: TokenBudgetConfig, scopeId: String?): AggregatedAIContext {
        val targetScopeId = scopeId ?: rootScopeId
        val hierarchy = mutableListOf<String>()
        var curr: ContextScope? = scopes[targetScopeId]

        val aggregatedVars = mutableMapOf<String, Any?>()
        while (curr != null) {
            hierarchy.add(0, curr.name)
            // Lower scopes override parent vars
            curr.variables.forEach { (k, v) ->
                if (!aggregatedVars.containsKey(k)) {
                    aggregatedVars[k] = v
                }
            }
            curr = curr.parentScopeId?.let { scopes[it] }
        }

        val validMemories = getMemorySlots(targetScopeId, config.minImportanceThreshold)

        val sb = StringBuilder()
        sb.appendLine("### ACTIVE CONTEXT SCOPE HIERARCHY")
        sb.appendLine("Path: ${hierarchy.joinToString(" -> ")}")
        sb.appendLine("\n### CONTEXT VARIABLES:")
        aggregatedVars.forEach { (k, v) ->
            sb.appendLine("- **$k**: $v")
        }

        if (config.includeMemorySlots && validMemories.isNotEmpty()) {
            sb.appendLine("\n### WORKING MEMORY (Prioritized by Importance):")
            validMemories.forEach { mem ->
                sb.appendLine("- [Imp: ${mem.importance}/10] **${mem.key}**: ${mem.value} (Tags: ${mem.tags.joinToString()})")
            }
        }

        val promptAddition = sb.toString()
        val estimatedTokens = (promptAddition.length / 4) + config.reservedSystemTokens

        return AggregatedAIContext(
            systemPromptAddition = promptAddition,
            structuredContext = aggregatedVars,
            relevantMemories = validMemories,
            estimatedTokens = estimatedTokens,
            scopeHierarchy = hierarchy
        )
    }

    override fun <R> transaction(scopeId: String?, block: (ContextEngine) -> R): R {
        val targetScopeId = scopeId ?: rootScopeId
        val backupSnapshot = getSnapshot(targetScopeId)
        return try {
            block(this)
        } catch (e: Exception) {
            restoreSnapshot(backupSnapshot, targetScopeId)
            throw e
        }
    }

    override fun reset() {
        scopes.clear()
        watchers.clear()
        diffHistory.clear()
        scopes[rootScopeId] = ContextScope(
            id = rootScopeId,
            name = "Global System Scope",
            metadata = ContextScopeMetadata(owner = "system", description = "Root system scope")
        )
    }

    private fun recordDiff(diff: ContextDiff) {
        diffHistory.add(diff)
        if (diffHistory.size > 200) {
            diffHistory.removeAt(0)
        }
    }

    private fun notifyWatchers(diff: ContextDiff, scope: ContextScope) {
        watchers.values.forEach { watcher ->
            if (watcher.scopeId == null || watcher.scopeId == diff.scopeId) {
                if (matchesPathPattern(watcher.pathPattern, diff.path)) {
                    try {
                        watcher.callback(diff, scope)
                    } catch (_: Exception) {}
                }
            }
        }
    }

    private fun matchesPathPattern(pattern: String, path: String): Boolean {
        if (pattern == "*" || pattern == "**" || pattern == path) return true
        val regex = pattern.replace(".", "\\.").replace("*", "[^.]+")
        return Pattern.compile("^$regex$").matcher(path).matches()
    }

    @Suppress("UNCHECKED_CAST")
    private fun getNestedValue(map: Map<String, Any?>, path: String): Any? {
        val parts = path.split(".")
        var current: Any? = map
        for (part in parts) {
            if (current !is Map<*, *>) return null
            current = (current as Map<String, Any?>)[part]
        }
        return current
    }

    @Suppress("UNCHECKED_CAST")
    private fun setNestedValue(map: ConcurrentHashMap<String, Any?>, path: String, value: Any?) {
        val parts = path.split(".")
        if (parts.size == 1) {
            if (value != null) map[parts[0]] = value else map.remove(parts[0])
            return
        }

        var current: MutableMap<String, Any?> = map
        for (i in 0 until parts.size - 1) {
            val part = parts[i]
            val next = current[part]
            if (next !is MutableMap<*, *>) {
                val newMap = ConcurrentHashMap<String, Any?>()
                current[part] = newMap
                current = newMap
            } else {
                current = next as MutableMap<String, Any?>
            }
        }
        val lastPart = parts.last()
        if (value != null) current[lastPart] = value else current.remove(lastPart)
    }

    @Suppress("UNCHECKED_CAST")
    private fun deleteNestedValue(map: MutableMap<String, Any?>, path: String): Boolean {
        val parts = path.split(".")
        if (parts.size == 1) {
            return map.remove(parts[0]) != null
        }
        var current: Any? = map
        for (i in 0 until parts.size - 1) {
            if (current !is Map<*, *>) return false
            current = (current as Map<String, Any?>)[parts[i]]
        }
        return (current as? MutableMap<String, Any?>)?.remove(parts.last()) != null
    }

    @Suppress("UNCHECKED_CAST")
    private fun deepCopyMap(original: Map<String, Any?>): Map<String, Any?> {
        val copy = mutableMapOf<String, Any?>()
        for ((k, v) in original) {
            copy[k] = when (v) {
                is Map<*, *> -> deepCopyMap(v as Map<String, Any?>)
                is List<*> -> v.toList()
                else -> v
            }
        }
        return copy
    }

    private data class WatcherEntry(
        val id: String,
        val pathPattern: String,
        val scopeId: String?,
        val callback: ContextWatcher
    )
}
